### PR TITLE
fix: bats test refactoring

### PR DESCRIPTION
**Pull request by adodon2go**
**What type of PR is this?**

**Which issue does this PR fix**:

**What does this PR do / Why do we need it**:

**If an issue # is not available please add repro steps and logs showing the issue**:

**Testing done on this change**:

**Automation added to e2e**:

**Will this break upgrades or downgrades?**

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: test/blackbox/anonymous_policy.bats [hunk 1]
```diff
@@ -1,8 +1,17 @@
-load helpers_pushpull
+load helpers_zot
+
+function verify_prerequisites {
+    if [ ! -f ${ZOT_PATH} ]; then
+        echo "you need to build ${ZOT_PATH} before running the tests" >&3
+        return 1
+    fi
+
+    return 0
```
Comment (andaaron) [inline test/blackbox/anonymous_policy.bats:9]: Don't we need to add a check for skopeo here, and in the other files using the skopeo command?

File: test/blackbox/annotations.bats [hunk 2]
```diff
@@ -1,21 +1,44 @@
...
+    if [ ! $(command -v jq) ]; then
+        echo "you need to install jq as a prerequisite to running the tests" >&3
+        return 1
+    fi
+
+    if [ ! $(command -v podman) ]; then
+        echo "you need to install podman as a prerequisite to running the tests" >&3
+        return 1
+    fi
+
```
Comment (andaaron) [inline test/blackbox/annotations.bats:23]: Don't we need to add checks or skopeo and stacker?

File: test/blackbox/pushpull.bats [hunk 3]
```diff
@@ -1,8 +1,27 @@
...
+
+    if [ ! $(command -v curl) ]; then
+        echo "you need to install curl as a prerequisite to running the tests" >&3
+        return 1
+    fi
+
+    if [ ! $(command -v jq) ]; then
+        echo "you need to install jq as a prerequisite to running the tests" >&3
+        return 1
+    fi
```
Comment (andaaron) [inline test/blackbox/pushpull.bats:17]: There's the missing tooling we should check for, helm, regctl, oras and so on?

File: test/blackbox/pushpull.bats [hunk 4]
```diff
@@ -146,6 +167,9 @@ function teardown_file() {
     local chart_version=$(awk '/version/{printf $2}' ${BATS_FILE_TMPDIR}/helm-charts/charts/zot/Chart.yaml)
     run helm pull oci://localhost:8080/zot-chart/zot --version ${chart_version}
     [ "$status" -eq 0 ]
+    local chart_version=$(awk '/version/{printf $2}' ${BATS_FILE_TMPDIR}/helm-charts/charts/zot/Chart.yaml)
```
Comment (andaaron) [inline test/blackbox/pushpull.bats:170]: Is is a duplicate line to 167.

File: test/blackbox/metadata.bats [hunk 5]
```diff
@@ -1,8 +1,27 @@
...
+    if [ ! $(command -v curl) ]; then
+        echo "you need to install curl as a prerequisite to running the tests" >&3
+        return 1
+    fi
+
+    if [ ! $(command -v jq) ]; then
+        echo "you need to install jq as a prerequisite to running the tests" >&3
+        return 1
+    fi
+
```
Comment (andaaron) [inline test/blackbox/metadata.bats:18]: We should add a check for skopeo.

File: test/blackbox/cve.bats [hunk 6]
```diff
@@ -1,9 +1,33 @@
...
+    if [ ! $(command -v curl) ]; then
+        echo "you need to install curl as a prerequisite to running the tests" >&3
+        return 1
+    fi
 
+    if [ ! $(command -v jq) ]; then
+        echo "you need to install jq as a prerequisite to running the tests" >&3
+        return 1
+    fi
+
```
Comment (andaaron) [inline test/blackbox/cve.bats:23]: We should check for skopeo.

File: test/blackbox/detect_manifest_collision.bats [hunk 7]
```diff
@@ -1,8 +1,17 @@
-load helpers_pushpull
+load helpers_zot
+
+function verify_prerequisites {
+    if [ ! -f ${ZOT_PATH} ]; then
+        echo "you need to build ${ZOT_PATH} before running the tests" >&3
+        return 1
+    fi
+
```
Comment (andaaron) [inline test/blackbox/detect_manifest_collision.bats:8]: We should check for skopeo and regctl.

File: test/blackbox/referrers.bats [hunk 8]
```diff
@@ -1,8 +1,27 @@
...
+    if [ ! $(command -v curl) ]; then
+        echo "you need to install curl as a prerequisite to running the tests" >&3
+        return 1
+    fi
+
+    if [ ! $(command -v jq) ] &>/dev/null; then
+        echo "you need to install jq as a prerequisite to running the tests" >&3
+        return 1
+    fi
+
```
Comment (andaaron) [inline test/blackbox/referrers.bats:18]: Check for skopeo and oras?

File: test/blackbox/pushpull_running_dedupe.bats [hunk 9]
```diff
@@ -1,8 +1,27 @@
...
+    if [ ! $(command -v curl) ]; then
+        echo "you need to install curl as a prerequisite to running the tests" >&3
+        return 1
+    fi
+
+    if [ ! $(command -v jq) ]; then
+        echo "you need to install jq as a prerequisite to running the tests" >&3
+        return 1
+    fi
+
```
Comment (andaaron) [inline test/blackbox/pushpull_running_dedupe.bats:18]: Check for skopeo, oras, helm, regctl.

File: test/blackbox/sync_replica_cluster.bats [hunk 10]
```diff
@@ -20,11 +20,11 @@ function setup_file() {
 
     cat >${zot_sync_one_config_file} <<EOF
 {
-    "distSpecVersion": "1.1.0",
+    "distSpecVersion": "1.1.0-dev",
     "storage": {
         "rootDirectory": "${zot_sync_one_root_dir}"
     },
-    "http": {
+    "http": {config
```
Comment (andaaron) [inline test/blackbox/sync_replica_cluster.bats:27]: Is this correct?

**Review by andaaron — COMMENTED** [review]
I agree with consolidating the helpers libraries, but I think it would be simpler to add all the checks for all required tooling there instead of individual .bats files. Why move it to the .bats files?

File: test/blackbox/helpers_pushpull.bash [hunk 11]
```diff
@@ -1,84 +0,0 @@
...
-TEST_DATA_DIR=${ROOT_DIR}/test/data/
-OS="${OS:-linux}"
-ARCH="${ARCH:-amd64}"
-ZOT_PATH=${ROOT_DIR}/bin/zot-${OS}-${ARCH}
-
-mkdir -p ${TEST_DATA_DIR}
-
-function verify_prerequisites {
-    if [ ! -f ${BATS_RUN_TMPDIR}/.firstrun ]; then
-        env | grep proxy >&3
```
Comment (andaaron) [inline test/blackbox/helpers_pushpull.bash:11]: Used to show the proxy, can you put it with `-i`

File: test/blackbox/annotations.bats (hunk 2 above)
Comment (adodon2go) [reply test/blackbox/annotations.bats:23]: This tools are already checked/installed if run through a Makefile target. If run as a standalone BATS script we need to make sure versions (ex: cosign.notation) are the same as the one installed by Makefile. This adds extra complexity that is probably not needed. We can add a note at the beginning of each file that states just that. Besides that there are some tests that did not pass locally anyway (that ones that requires crictl for ex or the sync one that syncs images from dockerhub - probably limitation of unpaid version)

**Review by andaaron — APPROVED** [review]

**Review by rchincha — APPROVED** [review]
lgtm

**Pull request merged**